torch/distributed/fsdp/flat_param.py

```diff
@@ -399,57 +442,70 @@ def shard(self, process_group: dist.ProcessGroup):
         unsharded flattened parameter's storage.
 
         Postcondition: ``self.flat_param`` is the sharded flattened parameter.
-        ``process_group``, ``rank``, and ``world_size`` attributes are set.
+        Shard metadata attributes are set for all sharding strategies.
+        ``process_group``, ``rank``, and ``world_size`` attributes are set if
+        using a sharded strategy.
 
         TODO (awgu): Once we retire ``FlattenParamsWrapper``, we should pass
         the process group directly to the ``FlatParamHandle`` constructor. For
         now, we decouple ``FlattenParamsWrapper` from a process group, but this
         makes the process-group-related attributes not necessarily defined.
         """
-        if not self.uses_sharded_strategy:
-            return
         flat_param = self.flat_param
         self.process_group = process_group
         self.rank = process_group.rank()
         self.world_size = process_group.size()
-        assert (
-            flat_param.storage_offset() == 0
-        ), "The `FlatParameter` is not the sole occupant of its storage"
-        orig_storage = flat_param.storage()
-        local_shard, numel_padded = FlatParamHandle._get_shard(
-            flat_param, self.rank, self.world_size
-        )
-        flat_param.set_(local_shard)  # type: ignore[call-overload]
-        self._init_shard_metadata(local_shard.numel(), numel_padded, self.rank)
-        if orig_storage.size() > 0:
-            orig_storage.resize_(0)
+        if not self.uses_sharded_strategy:
+            self._init_shard_metadata(0, 0, flat_param.numel() - 1)
```

Comment:
what significance does flat_param.numel - 1 have for unsharded strategy?

Reply:
This line is not strictly needed and was legacy before I decided to have `_use_sharded_views()` to route to `_use_unsharded_views()` for `NO_SHARD`. However, it is not bad to keep the attribute existences more consistent across sharded and non-sharded strategies.

The `flat_param.numel() - 1` is the inclusive end of the rank's "shard", i.e. this rank owns `[0, flat_param.numel() - 1]`, which is the entire parameter since the strategy is `NO_SHARD`.